Impossible to use multiple Lua scripts with exporter

**Describe the problem**
Hello, as I can see from the code and experimentally, there is no way to define couple of Lua scripts to evaluate by exporter. Only one is working in the end. I 

If there is no critical requirement, is it possible improve exporter to handle multiple scripts?
Or add information to Readme about this limitations.

**What version of redis_exporter are you running?**
`Redis Metrics Exporter v1.37.0    build date: 2022-03-18-01:20:01    sha1: a1c28b775760f2f00fce07a24db7fd4e83c26b9f    Go: go1.17.8    GOOS: linux    GOARCH: arm64`
* [ ] 0.3x.x
* [x] 1.x.x


**Running the exporter**
```sh
redis_exporter -redis.addr redis://localhost:6379 -script script1.lua -script script2.lua
```
Only last script is working.

**Screenshots**
None


**Additional context**
None

Sure, thanks, will try to create a PR around next week.

I'm thinking of add new option like `-script-folder` to pass whole folder with Lua scripts.
And maybe allow multiple definition of current `-script` like [here](https://go.dev/play/p/Ig5sm7jA14), but it will probably break env variable `REDIS_EXPORTER_SCRIPT` support.

Also, as an option, we can just split what's passed to `-script` or env var by `,` for example, which allow to pass multiple scripts separated by `,`.

What do you think?

--- a/exporter/exporter.go
+++ b/exporter/exporter.go
@@ -59,7 +59,7 @@ type Options struct {
 	CheckKeyGroups        string
 	MaxDistinctKeyGroups  int64
 	CountKeys             string
-	LuaScript             []byte
+	LuaScript             [][]byte
 	ClientCertFile        string
 	ClientKeyFile         string
 	CaCertFile            string
@@ -648,8 +648,10 @@ func (e *Exporter) scrapeRedisHost(ch chan<- prometheus.Metric) error {
 	}
 
 	if len(e.options.LuaScript) > 0 {
-		if err := e.extractLuaScriptMetrics(ch, c); err != nil {
-			return err
+		for _, script := range e.options.LuaScript {
+			if err := e.extractLuaScriptMetrics(ch, c, script); err != nil {
+				return err
+			}
 		}
 	}
 

--- a/exporter/exporter.go
+++ b/exporter/exporter.go
@@ -59,7 +59,7 @@ type Options struct {
 	CheckKeyGroups        string
 	MaxDistinctKeyGroups  int64
 	CountKeys             string
-	LuaScript             [][]byte
+	LuaScript             map[string][]byte
 	ClientCertFile        string
 	ClientKeyFile         string
 	CaCertFile            string
@@ -374,7 +374,7 @@ func NewRedisExporter(redisURI string, opts Options) (*Exporter, error) {
 		"master_link_up":                               {txt: "Master link status on Redis slave", lbls: []string{"master_host", "master_port"}},
 		"master_sync_in_progress":                      {txt: "Master sync in progress", lbls: []string{"master_host", "master_port"}},
 		"number_of_distinct_key_groups":                {txt: `Number of distinct key groups`, lbls: []string{"db"}},
-		"script_values":                                {txt: "Values returned by the collect script", lbls: []string{"key"}},
+		"script_values":                                {txt: "Values returned by the collect script", lbls: []string{"key", "filename"}},
 		"sentinel_master_ok_sentinels":                 {txt: "The number of okay sentinels monitoring this master", lbls: []string{"master_name", "master_address"}},
 		"sentinel_master_ok_slaves":                    {txt: "The number of okay slaves of the master", lbls: []string{"master_name", "master_address"}},
 		"sentinel_master_sentinels":                    {txt: "The number of sentinels monitoring this master", lbls: []string{"master_name", "master_address"}},
@@ -648,8 +648,8 @@ func (e *Exporter) scrapeRedisHost(ch chan<- prometheus.Metric) error {
 	}
 
 	if len(e.options.LuaScript) > 0 {
-		for _, script := range e.options.LuaScript {
-			if err := e.extractLuaScriptMetrics(ch, c, script); err != nil {
+		for filename, script := range e.options.LuaScript {
+			if err := e.extractLuaScriptMetrics(ch, c, filename, script); err != nil {
 				return err
 			}
 		}

--- a/exporter/exporter.go
+++ b/exporter/exporter.go
@@ -374,6 +374,7 @@ func NewRedisExporter(redisURI string, opts Options) (*Exporter, error) {
 		"master_link_up":                               {txt: "Master link status on Redis slave", lbls: []string{"master_host", "master_port"}},
 		"master_sync_in_progress":                      {txt: "Master sync in progress", lbls: []string{"master_host", "master_port"}},
 		"number_of_distinct_key_groups":                {txt: `Number of distinct key groups`, lbls: []string{"db"}},
+		"script_result":                                {txt: "Result of the collect script evaluation", lbls: []string{"filename"}},
 		"script_values":                                {txt: "Values returned by the collect script", lbls: []string{"key", "filename"}},
 		"sentinel_master_ok_sentinels":                 {txt: "The number of okay sentinels monitoring this master", lbls: []string{"master_name", "master_address"}},
 		"sentinel_master_ok_slaves":                    {txt: "The number of okay slaves of the master", lbls: []string{"master_name", "master_address"}},

--- a/exporter/http_test.go
+++ b/exporter/http_test.go
@@ -83,8 +83,8 @@ func TestHTTPScrapeMetricsEndpoints(t *testing.T) {
 			options := Options{
 				Namespace: "test",
 				Password:  tst.pwd,
-				LuaScript: [][]byte{
-					[]byte(`return {"a", "11", "b", "12", "c", "13"}`),
+				LuaScript: map[string][]byte{
+					"test.lua": []byte(`return {"a", "11", "b", "12", "c", "13"}`),
 				},
 				Registry: prometheus.NewRegistry(),
 			}

--- a/exporter/http_test.go
+++ b/exporter/http_test.go
@@ -83,8 +83,10 @@ func TestHTTPScrapeMetricsEndpoints(t *testing.T) {
 			options := Options{
 				Namespace: "test",
 				Password:  tst.pwd,
-				LuaScript: []byte(`return {"a", "11", "b", "12", "c", "13"}`),
-				Registry:  prometheus.NewRegistry(),
+				LuaScript: [][]byte{
+					[]byte(`return {"a", "11", "b", "12", "c", "13"}`),
+				},
+				Registry: prometheus.NewRegistry(),
 			}
 
 			options.CheckSingleKeys = tst.csk

--- a/exporter/lua.go
+++ b/exporter/lua.go
@@ -8,9 +8,9 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
-func (e *Exporter) extractLuaScriptMetrics(ch chan<- prometheus.Metric, c redis.Conn) error {
+func (e *Exporter) extractLuaScriptMetrics(ch chan<- prometheus.Metric, c redis.Conn, script []byte) error {
 	log.Debug("Evaluating e.options.LuaScript")
-	kv, err := redis.StringMap(doRedisCmd(c, "EVAL", e.options.LuaScript, 0, 0))
+	kv, err := redis.StringMap(doRedisCmd(c, "EVAL", script, 0, 0))
 	if err != nil {
 		log.Errorf("LuaScript error: %v", err)
 		return err

--- a/exporter/lua.go
+++ b/exporter/lua.go
@@ -27,6 +27,7 @@ func (e *Exporter) extractLuaScriptMetrics(ch chan<- prometheus.Metric, c redis.
 		val, err := strconv.ParseFloat(stringVal, 64)
 		if err != nil {
 			log.Errorf("Error parsing lua script results, err: %s", err)
+			e.registerConstMetricGauge(ch, "script_result", 0, filename)
 			return err
 		}
 		e.registerConstMetricGauge(ch, "script_values", val, key, filename)

--- a/exporter/lua.go
+++ b/exporter/lua.go
@@ -8,8 +8,8 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
-func (e *Exporter) extractLuaScriptMetrics(ch chan<- prometheus.Metric, c redis.Conn, script []byte) error {
-	log.Debug("Evaluating e.options.LuaScript")
+func (e *Exporter) extractLuaScriptMetrics(ch chan<- prometheus.Metric, c redis.Conn, filename string, script []byte) error {
+	log.Debugf("Evaluating e.options.LuaScript: %s", filename)
 	kv, err := redis.StringMap(doRedisCmd(c, "EVAL", script, 0, 0))
 	if err != nil {
 		log.Errorf("LuaScript error: %v", err)
@@ -27,7 +27,7 @@ func (e *Exporter) extractLuaScriptMetrics(ch chan<- prometheus.Metric, c redis.
 			log.Errorf("Error parsing lua script results, err: %s", err)
 			return err
 		}
-		e.registerConstMetricGauge(ch, "script_values", val, key)
+		e.registerConstMetricGauge(ch, "script_values", val, key, filename)
 	}
 	return nil
 }

--- a/exporter/lua.go
+++ b/exporter/lua.go
@@ -13,11 +13,13 @@ func (e *Exporter) extractLuaScriptMetrics(ch chan<- prometheus.Metric, c redis.
 	kv, err := redis.StringMap(doRedisCmd(c, "EVAL", script, 0, 0))
 	if err != nil {
 		log.Errorf("LuaScript error: %v", err)
+		e.registerConstMetricGauge(ch, "script_result", 0, filename)
 		return err
 	}
 
 	if len(kv) == 0 {
 		log.Debugf("Lua script returned no results")
+		e.registerConstMetricGauge(ch, "script_result", 2, filename)
 		return nil
 	}
 
@@ -29,5 +31,6 @@ func (e *Exporter) extractLuaScriptMetrics(ch chan<- prometheus.Metric, c redis.
 		}
 		e.registerConstMetricGauge(ch, "script_values", val, key, filename)
 	}
+	e.registerConstMetricGauge(ch, "script_result", 1, filename)
 	return nil
 }

--- a/exporter/lua_test.go
+++ b/exporter/lua_test.go
@@ -54,7 +54,7 @@ func TestLuaScript(t *testing.T) {
 				os.Getenv("TEST_REDIS_URI"),
 				Options{
 					Namespace: "test", Registry: prometheus.NewRegistry(),
-					LuaScript: [][]byte{[]byte(tst.Script)},
+					LuaScript: map[string][]byte{"test.lua": []byte(tst.Script)},
 				})
 			ts := httptest.NewServer(e)
 			defer ts.Close()

--- a/exporter/lua_test.go
+++ b/exporter/lua_test.go
@@ -54,7 +54,7 @@ func TestLuaScript(t *testing.T) {
 				os.Getenv("TEST_REDIS_URI"),
 				Options{
 					Namespace: "test", Registry: prometheus.NewRegistry(),
-					LuaScript: []byte(tst.Script),
+					LuaScript: [][]byte{[]byte(tst.Script)},
 				})
 			ts := httptest.NewServer(e)
 			defer ts.Close()

--- a/exporter/lua_test.go
+++ b/exporter/lua_test.go
@@ -20,33 +20,34 @@ func TestLuaScript(t *testing.T) {
 		{
 			Name:         "ok1",
 			Script:       `return {"a", "11", "b", "12", "c", "13"}`,
-			ExpectedKeys: 3,
-			Wants:        []string{`test_exporter_last_scrape_error{err=""} 0`, `test_script_values{key="a"} 11`, `test_script_values{key="b"} 12`},
+			ExpectedKeys: 4,
+			Wants:        []string{`test_exporter_last_scrape_error{err=""} 0`, `test_script_values{filename="test.lua",key="a"} 11`, `test_script_values{filename="test.lua",key="b"} 12`, `test_script_values{filename="test.lua",key="c"} 13`, `test_script_result{filename="test.lua"} 1`},
 		},
 		{
 			Name:         "ok2",
 			Script:       `return {"key1", "6389"}`,
-			ExpectedKeys: 1,
-			Wants:        []string{`test_exporter_last_scrape_error{err=""} 0`, `test_script_values{key="key1"} 6389`},
+			ExpectedKeys: 4,
+			Wants:        []string{`test_exporter_last_scrape_error{err=""} 0`, `test_script_values{filename="test.lua",key="key1"} 6389`, `test_script_result{filename="test.lua"} 1`},
 		},
 		{
 			Name:         "ok3",
 			Script:       `return {} `,
-			ExpectedKeys: 0,
+			ExpectedKeys: 1,
+			Wants:        []string{`test_script_result{filename="test.lua"} 2`},
 		},
 		{
 			Name:          "borked1",
 			Script:        `return {"key1"   BROKEN `,
-			ExpectedKeys:  0,
+			ExpectedKeys:  1,
 			ExpectedError: true,
-			Wants:         []string{`test_exporter_last_scrape_error{err="ERR Error compiling script`},
+			Wants:         []string{`test_exporter_last_scrape_error{err="ERR Error compiling script`, `test_script_result{filename="test.lua"} 0`},
 		},
 		{
 			Name:          "borked2",
 			Script:        `return {"key1", "abc"}`,
-			ExpectedKeys:  0,
+			ExpectedKeys:  1,
 			ExpectedError: true,
-			Wants:         []string{`test_exporter_last_scrape_error{err="strconv.ParseFloat: parsing \"abc\": invalid syntax"} 1`},
+			Wants:         []string{`test_exporter_last_scrape_error{err="strconv.ParseFloat: parsing \"abc\": invalid syntax"} 1`, `test_script_result{filename="test.lua"} 0`},
 		},
 	} {
 		t.Run(tst.Name, func(t *testing.T) {

--- a/exporter/pwd_file_test.go
+++ b/exporter/pwd_file_test.go
@@ -103,8 +103,10 @@ func TestHTTPScrapeWithPasswordFile(t *testing.T) {
 		options := Options{
 			Namespace:   "test",
 			PasswordMap: passwordMap,
-			LuaScript:   []byte(`return {"a", "11", "b", "12", "c", "13"}`),
-			Registry:    prometheus.NewRegistry(),
+			LuaScript: [][]byte{
+				[]byte(`return {"a", "11", "b", "12", "c", "13"}`),
+			},
+			Registry: prometheus.NewRegistry(),
 		}
 		t.Run(tst.name, func(t *testing.T) {
 			e, _ := NewRedisExporter(tst.addr, options)

--- a/exporter/pwd_file_test.go
+++ b/exporter/pwd_file_test.go
@@ -103,8 +103,8 @@ func TestHTTPScrapeWithPasswordFile(t *testing.T) {
 		options := Options{
 			Namespace:   "test",
 			PasswordMap: passwordMap,
-			LuaScript: [][]byte{
-				[]byte(`return {"a", "11", "b", "12", "c", "13"}`),
+			LuaScript: map[string][]byte{
+				"test.lua": []byte(`return {"a", "11", "b", "12", "c", "13"}`),
 			},
 			Registry: prometheus.NewRegistry(),
 		}

--- a/main.go
+++ b/main.go
@@ -7,6 +7,7 @@ import (
 	"os"
 	"runtime"
 	"strconv"
+	"strings"
 	"time"
 
 	"github.com/prometheus/client_golang/prometheus"
@@ -65,7 +66,7 @@ func main() {
 		checkSingleStreams   = flag.String("check-single-streams", getEnv("REDIS_EXPORTER_CHECK_SINGLE_STREAMS", ""), "Comma separated list of single streams to export info about streams, groups and consumers")
 		countKeys            = flag.String("count-keys", getEnv("REDIS_EXPORTER_COUNT_KEYS", ""), "Comma separated list of patterns to count (eg: 'db0=production_*,db3=sessions:*'), searched for with SCAN")
 		checkKeysBatchSize   = flag.Int64("check-keys-batch-size", getEnvInt64("REDIS_EXPORTER_CHECK_KEYS_BATCH_SIZE", 1000), "Approximate number of keys to process in each execution, larger value speeds up scanning.\nWARNING: Still Redis is a single-threaded app, huge COUNT can affect production environment.")
-		scriptPath           = flag.String("script", getEnv("REDIS_EXPORTER_SCRIPT", ""), "Path to Lua Redis script for collecting extra metrics")
+		scriptPath           = flag.String("script", getEnv("REDIS_EXPORTER_SCRIPT", ""), "Path to Lua Redis scripts for collecting extra metrics, separated by ,")
 		listenAddress        = flag.String("web.listen-address", getEnv("REDIS_EXPORTER_WEB_LISTEN_ADDRESS", ":9121"), "Address to listen on for web interface and telemetry.")
 		metricPath           = flag.String("web.telemetry-path", getEnv("REDIS_EXPORTER_WEB_TELEMETRY_PATH", "/metrics"), "Path under which to expose metrics.")
 		logFormat            = flag.String("log-format", getEnv("REDIS_EXPORTER_LOG_FORMAT", "txt"), "Log format, valid options are txt and json")
@@ -133,10 +134,14 @@ func main() {
 		}
 	}
 
-	var ls []byte
+	var ls [][]byte
 	if *scriptPath != "" {
-		if ls, err = ioutil.ReadFile(*scriptPath); err != nil {
-			log.Fatalf("Error loading script file %s    err: %s", *scriptPath, err)
+		scripts := strings.Split(*scriptPath, ",")
+		ls = make([][]byte, len(scripts))
+		for i, script := range scripts {
+			if ls[i], err = ioutil.ReadFile(script); err != nil {
+				log.Fatalf("Error loading script file %s    err: %s", script, err)
+			}
 		}
 	}
 

--- a/main.go
+++ b/main.go
@@ -134,12 +134,12 @@ func main() {
 		}
 	}
 
-	var ls [][]byte
+	var ls map[string][]byte
 	if *scriptPath != "" {
 		scripts := strings.Split(*scriptPath, ",")
-		ls = make([][]byte, len(scripts))
-		for i, script := range scripts {
-			if ls[i], err = ioutil.ReadFile(script); err != nil {
+		ls = make(map[string][]byte, len(scripts))
+		for _, script := range scripts {
+			if ls[script], err = ioutil.ReadFile(script); err != nil {
 				log.Fatalf("Error loading script file %s    err: %s", script, err)
 			}
 		}

